src/auth/authApp.js

```diff
@@ -84,23 +82,123 @@ export function redirectToSignInWithAuthRequest(authRequest: string = makeAuthRe
+      }
+    }, 100)
   }
+  inputPromptTracker.addEventListener('blur', inputBlurredFunc, { once: true, capture: true })
+  setTimeout(() => inputPromptTracker.removeEventListener('blur', inputBlurredFunc), 200)
+  // Flow complains without this check.
+  if (document.body) document.body.appendChild(inputPromptTracker)
+  inputPromptTracker.focus()
 
-  function unsupportedBrowserCallback() {
-    // Safari is unsupported by protocolCheck
-    Logger.warn('can not detect custom protocols on this browser')
-    window.location = protocolURI
+  // Detect if document.visibility is immediately changed which is a strong 
+  // indication that the protocol handler is working. We don't know for sure and 
+  // can't predict future browser changes, so only increase the redirect timeout.
+  // This reduces the probability of a false-negative (where local auth works, but 
+  // the original page was redirect to web auth because something took too long),
+  const pageVisibilityChanged = () => {
+    if (document.hidden && redirectToWebAuthTimer) {
+      Logger.info('Detected immediate page visibility change (protocol handler probably working).')
+      startWebAuthRedirectTimer(3000)
+    }
   }
+  document.addEventListener('visibilitychange', pageVisibilityChanged, { once: true, capture: true })
+  setTimeout(() => document.removeEventListener('visibilitychange', pageVisibilityChanged), 500)
 
-  protocolCheck(protocolURI, failCallback, successCallback, unsupportedBrowserCallback)
+
+  // Listen for the custom protocol echo reply via localStorage update event.
+  window.addEventListener('storage', function replyEventListener(event) {
+    if (event.key === echoReplyKey && window.localStorage.getItem(echoReplyKey) === 'success') {
+      // Custom protocol worked, cancel the web auth redirect timer.
+      cancelWebAuthRedirectTimer()
+      inputPromptTracker.removeEventListener('blur', inputBlurredFunc)
+      Logger.info('Protocol handler detected.')
+      // Clean up event listener and localStorage.
+      window.removeEventListener('storage', replyEventListener)
+      cleanUpLocalStorage()
+    }
+  }, false)
+
+  // Use iframe technique for launching the protocol URI rather than setting `window.location`.
+  // This method prevents browsers like Safari, Opera, Firefox from showing error prompts
+  // about unknown protocol handler when app is not installed, and avoids an empty
+  // browser tab when the app is installed. 
+  Logger.info('Attempting protocol launch via iframe injection.')
+  const locationSrc = `${BLOCKSTACK_HANDLER}:${authRequest}&echo=${echoReplyID}`
+  const iframe = document.createElement('iframe')
+  iframe.style.all = 'initial'
+  iframe.style.display = 'none'
+  iframe.src = locationSrc
+  // Flow complains without this check.
+  if (document.body) {
+    document.body.appendChild(iframe)
+  } else {
+    Logger.error('document.body is null when attempting iframe injection for protoocol URI launch')
+  }
```

Comment:
Can this be refactored to look somewhat like `checkedBrowserLaunch(protocolURI, failCallback)` ?

It'd make it a bit more clear later on if we want to do something other than just redirect to browser.blockstack.org.

Reply:
Good idea, I will do that.

Reply:
Refactored the protocol detection code out of `redirectToSignInWithAuthRequest` function and into new `detectProtocolLaunch` function, with callbacks for handlers similar to how it was before.

Reply:
@kantai want to take a look at see if we can resolve this comment?

Reply:
Yep, looks good, thanks for the refactor!